### ISSUE TITLE
Best practices with flask

### ISSUE THREAD

**Issue by MSchutt** [ClickHouse/clickhouse-connect#73]
Good Evening!
Thank you for the package. It seems very well written and everything works so far.
I am currently working on a project with flask and I am unsure on how to properly integrate.
I have 3 possible implementation paths:
* Connection Pooling
* Create a new client on each request
* Create a singleton for all requests

Maybe we can discuss and then build another package upon it or write a small section in the readme as I believe this is a common use case.
Please let me know what you think - Thank you!

**Comment by genzgd** [comment]
I wouldn't recommend creating a new client on each request, as there's some overhead on startup (in particular, retrieving ClickHouse server information and settings), and you would lose the advantage of the HTTP and TCP keep alive settings for the underlying HTTP and TCP connections.  And unless you are changing overall client settings, each request is stateless and thread safe.

Currently all clients share a hard coded requests/urllib3 connection pool, using this code from the httpclient.py file:

```python
# Create a single HttpAdapter that will be shared by all client sessions.  This is intended to make
# the client as thread safe as possible while sharing a single connection pool.  For the same reason we
# don't call the Session.close() method from the client so the connection pool remains available
http_adapter = KeepAliveAdapter(pool_connections=4, pool_maxsize=8, max_retries=0)
atexit.register(http_adapter.close)
```

Because that's hard coded at a maximum of 8 connections, it seems like making a singleton might artificially limit the scalability of your Flask application.  Although I know this was a quick and dirty approach, you're the first person to (correctly) suggest that in some cases something more flexible is required.

I've not done anything extensive with Flask, but I can see a couple of options.  One is to make that connection pool configurable on a global level, the other is to allow passing an optional HTTPAdapter (or the configuration to make a new one) in to the HttpClient constructor.  Do you want to play around with those options and see what meets your use case?   Either approach should be fairly easy to implement. 

I'm also open to a small additional Flask package, but I'm unlikely to get to something like that any time soon, and I don't have any sense of how complex such a project would be.

**Comment by MSchutt** [comment]
Thank you for your input. I have played around and tried various options. Internal pooling of the connection (so creating multiple client object) does not make too much sense and I have faced some performance problems when creating a new instance for every request (which makes sense as it fetches various infos about the instance).

I think that it would be best to either expose the KeepAliveAdapter configuration or (as you mentioned) be able to pass an optional HTTPAdapter (would probably be the most clean solution in my opinion) and then use the connection as a singleton throughout the project. What do you think?

**Comment by genzgd** [comment]
I've the optional HTTPAdapter to the HttpClient constructor, so you should be able to use one properly configured adapter per target host (or otherwise mix and match clients and adapters keeping in mind the shared connection pool behind the adapter).

**Comment by MSchutt** [comment]
Awesome. Thank you - passing a custom HTTPAdapater works perfectly now. I am now facing a problem that for long queries I sometimes get the error:
`Code: 373. DB::Exception: Session is locked by a concurrent client. (SESSION_IS_LOCKED) (version 22.8.2.11 (official build))` which of course makes sense, as all requests share the same session_id. Is there a way to have different session_ids for each requests/instance of the connection pool?

**Comment by genzgd** [comment]
Yes, as you've figured out, ClickHouse won't allow queries to execute concurrently with the same session id.  The session id is optional -- also there's a setting to disable creating an automatic session_id when constructing the client.  You can change the default when initializing the driver as follows:

```python
import clickhouse_connect
clickhouse_connect.common.set_setting('autogenerate_session_id', False)
```

So depending on your use case, if you don't actually need a persistent session, that would solve the problem.  In those instances where a session is required (persisting settings creating by the `SET` command for example), I can't think of any great options.  You could construct a client per session (sharing the connection pool but not the session id), or pass the desired session id with every query (as an element of the settings dictionary), but I'm open to better ideas.

**Comment by MSchutt** [comment]
Thank you once again.
I tried the snippet you pasted (using Python 3.9.10):
```
import clickhouse_connect 
clickhouse_connect.common.set_setting('autogenerate_session_id', False)
``` 

But I receive the error:
```
…/site-packages/clickhouse_connect/common.py", line 38, in set_setting
    setting.value = value
AttributeError: can't set attribute
```

The `CommonSetting` class is implemented as a subclass of `NamedTuple` which is immutable:

```
class CommonSetting(NamedTuple):
    name: str
    options: Sequence[Any]
    default: Any
    value: Optional[Any] = None
```

I think that it would be sufficient to implement a getter and setter for the value property. What do you think?

Thanks again for your advice. I think that it would be nice to offer the possibility to pass a session_id per query (that should be possible by passing the `settings` dictionary to the query, right?

**Comment by genzgd** [comment]
ack, yeah that's a bug with the immutable named tuple.  It should probably just be a regular class.

And yes, you should be able to use the setting dictionary to pass a `session_id` for each query (note that ClickHouse will automatically generate a session id if none is provided)

**Comment by genzgd** [comment]
The immutable settings issue should be fixed by version 0.4.5

**Comment by MSchutt** [comment]
Awesome everything is working just fine. Thank you so much for your support!
